plugins/record-hook/src/RecordHook.ts

```diff
@@ -32,33 +37,59 @@ export const RecordHook = async (
   )
 }
 
+export interface BulkRecordHookOptions {
+  chunkSize?: number
+  parallel?: number
+  debug?: boolean
+}
+
 export const BulkRecordHook = async (
   event: FlatfileEvent,
   handler: (
     records: FlatfileRecord[],
     event?: FlatfileEvent
   ) => any | Promise<any>,
-  options: { chunkSize?: number; parallel?: number; debug?: boolean } = {}
+  options: BulkRecordHookOptions = {}
 ) => {
+  const { versionId } = event.context
+
+  const fetchData = async () => {
+    const data = await event.data
+    return data.records && data.records.length
+      ? prepareXRecords(data.records)
+      : undefined
+  }
+
   try {
     const batch = await event.cache.init<FlatfileRecords<any>>(
       'records',
-      async () => {
-        const data = await event.data
-        return prepareXRecords(data.records)
-      }
+      fetchData
     )
 
-    if (!batch || batch.records.length === 0) return
+    if (!batch || batch.records.length === 0) {
+      if (options.debug) {
+        console.log('No records to process')
+      }
+      await api.commits.complete(versionId)
+      return
+    }
 
-    // run client defined data hooks
+    // Execute client-defined data hooks
     await asyncBatch(batch.records, handler, options, event)
 
     event.afterAll(async () => {
-      const batch = event.cache.get<FlatfileRecords<any>>('records')
+      const cachedBatch = event.cache.get<FlatfileRecords<any>>('records')
+      const modifiedRecords = cachedBatch.records.filter(hasRecordChanges)
+      if (!modifiedRecords || modifiedRecords.length === 0) {
+        if (options.debug) {
+          console.log('No records modified')
+        }
+        await api.commits.complete(versionId)
+        return
+      }
 
-      const recordsUpdates = new RecordTranslater<FlatfileRecord>(
-        batch.records
+      const recordsUpdates = new RecordTranslator<FlatfileRecord>(
+        modifiedRecords
       ).toXRecords()
 
       try {
```

Comment:
(next line) `event.update()` will complete the commit, so no need to manually complete it here.